File: frontend/src/lib/external/plantnet.js
// Cliente de Pl@ntNet (identificación de especies por foto, en tiempo real)
const PLANTNET_API = 'https://my-api.plantnet.org/v2';

export function getApiKey() {
  return process.env.PLANTNET_API_KEY || '';
}

// Identifica una o varias imágenes; devuelve las N mejores coincidencias
export async function identify(images, { nbResults = 5 } = {}) {
  const apiKey = getApiKey();
  if (!apiKey) {
    throw new Error('PLANTNET_API_KEY no configurada');
  }
  const params = new URLSearchParams({
    'api-key': apiKey,
    'nb-results': String(nbResults),
  });

  const form = new FormData();
  for (const img of images) {
    const blob = new Blob([img.buffer], { type: img.contentType });
    form.append('images', blob, img.filename);
  }

  const res = await fetch(`${PLANTNET_API}/identify/all?${params}`, {
    method: 'POST',
    body: form,
  });
  if (res.status === 401 || res.status === 403) {
    throw new Error('API key de Pl@ntNet inválida o sin permisos');
  }
  if (!res.ok) {
    const body = await res.text().catch(() => '');
    throw new Error(`Pl@ntNet HTTP ${res.status}: ${body.slice(0, 200)}`);
  }

  const data = await res.json();
  const results = data.results || [];
  return results.slice(0, nbResults).map((r) => ({
    score: r.score ?? null,
    scientificName: r.species?.scientificNameWithoutAuthor || r.species?.scientificName || null,
    commonName: (r.species?.commonNames && r.species.commonNames[0]) || null,
    genus: r.species?.genus?.scientificName || null,
    family: r.species?.family?.scientificName || null,
  }));
}

File: backend/app/api/routers/plantnet.py
import os
import httpx
from fastapi import APIRouter, HTTPException, UploadFile, File, Request
from .users import get_current_user_id

router = APIRouter(
    prefix="/plantnet",
    tags=["plantnet"]
)

PLANTNET_API = 'https://my-api.plantnet.org/v2'

@router.post("/identify")
async def identify_plant(
    request: Request,
    image: UploadFile = File(...)
):
    # Protect with auth middleware equivalent
    user_id = get_current_user_id(request)
    
    api_key = os.getenv("PLANTNET_API_KEY")
    if not api_key:
        raise HTTPException(status_code=503, detail="Pl@ntNet no está configurado en este servidor")
        
    if not image.content_type or not image.content_type.startswith("image/"):
        raise HTTPException(status_code=400, detail="El archivo debe ser una imagen")
        
    contents = await image.read()
    MAX_SIZE = 10 * 1024 * 1024
    if len(contents) > MAX_SIZE:
        raise HTTPException(status_code=400, detail="La imagen supera el tamaño máximo (10 MB)")
        
    filename = image.filename or f"observacion.{image.content_type.split('/')[1] if '/' in image.content_type else 'jpg'}"
    
    files = {
        'images': (filename, contents, image.content_type)
    }
    params = {
        'api-key': api_key,
        'nb-results': '5'
    }
    
    async with httpx.AsyncClient() as client:
        try:
            res = await client.post(f"{PLANTNET_API}/identify/all", params=params, files=files)
            if res.status_code in (401, 403):
                raise HTTPException(status_code=401, detail="API key de Pl@ntNet inválida o sin permisos")
            if not res.is_success:
                body_text = res.text[:200]
                raise HTTPException(status_code=500, detail=f"Pl@ntNet HTTP {res.status_code}: {body_text}")
                
            data = res.json()
            results_raw = data.get("results", [])
            results = []
            
            for r in results_raw[:5]:
                species = r.get("species", {})
                scientificName = species.get("scientificNameWithoutAuthor") or species.get("scientificName")
                commonNames = species.get("commonNames", [])
                commonName = commonNames[0] if commonNames else None
                genus = species.get("genus", {}).get("scientificName")
                family = species.get("family", {}).get("scientificName")
                
                results.append({
                    "score": r.get("score"),
                    "scientificName": scientificName,
                    "commonName": commonName,
                    "genus": genus,
                    "family": family
                })
            
            return {"results": results}
            
        except httpx.RequestError as e:
            raise HTTPException(status_code=500, detail=f"Error al identificar la especie: {str(e)}")
